### PR TITLE
FDP-370 Fix out of sync invocation counter with dynamic IP address

### PR DESCRIPTION
**Pull request by sanderv**
Fixes initialization of out of sync invocation counter with DLMS devices that have a dynamic IP address and get a session from the session provider for the communication provider in use with GPRS.

### CODE REVIEW

**Comment by bvdzwet** [comment]
retest this please

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java [hunk 1]
```diff
@@ -115,7 +115,7 @@ public void run() {
 
   private DlmsDevice findDevice() {
     try {
-      return this.domainHelperService.findDlmsDevice(this.deviceIdentification, this.ipAddress);
+      return this.domainHelperService.findDlmsDevice(this.deviceIdentification);
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java:118]: If `this.ipAddress` is not used here anymore, can it be removed?

Comment (bvdzwet) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java:118]: It is still used in the code (in `checkState` and from the `RecoverKeyProcessInitiator`) though you are right in the sense that the IP address from this field is no longer used to set up the device connection.
I will update the state validation and initiation from the `RecoverKeyProcessInitiator` in order to get rid of the `ipAddress` field in `RecoverKeyProcess`.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java [hunk 2]
```diff
@@ -150,6 +150,14 @@ private boolean canConnectUsingNewKeys(final DlmsDevice device) {
...
       }
 
+      if (device.isIpAddressIsStatic()) {
+        device.setIpAddress(this.messageMetadata.getIpAddress());
+      } else {
+        final String ipAddressFromSessionProvider =
+            this.domainHelperService.getDeviceIpAddressFromSessionProvider(device);
+        device.setIpAddress(ipAddressFromSessionProvider);
+      }
+
```
[the remark this answers is not in the document]
Comment (bvdzwet) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/threads/RecoverKeyProcess.java:160]: I will refactor this to a method called `setIpAddressFromMessageMetadataOrSessionProvider`. Although not as bad as with a finder method like `findDlmsDevice`, the side effects of reaching out to the session provider might be hidden a little too much with the name being just `setIpAddress`.

**Review by stefanermens — APPROVED** [review]

**Pull request merged**